File: src/main/java/com/ptaf/ai/audit/AiGenerationAuditLogger.java
package com.ptaf.ai.audit;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.security.MessageDigest;
import java.util.HexFormat;

/**
 * Append-only JSONL logger for AI generation audit records.
 */
public final class AiGenerationAuditLogger {
    private static final ObjectMapper JSON = new ObjectMapper();

    public record WriteResult(boolean written, String outputPath, String warningMessage) {
    }

    public WriteResult append(Path projectRoot, boolean enabled, String outputPath, AiGenerationAuditRecord record) {
        if (!enabled) {
            return new WriteResult(false, outputPath, null);
        }
        try {
            Path file = projectRoot.resolve(outputPath).normalize();
            Path parent = file.getParent();
            if (parent != null) {
                Files.createDirectories(parent);
            }
            Files.writeString(
                    file,
                    JSON.writeValueAsString(record) + "\n",
                    StandardCharsets.UTF_8,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND
            );
            return new WriteResult(true, file.toString(), null);
        } catch (Exception e) {
            return new WriteResult(false, outputPath, "Warning: audit log could not be written");
        }
    }

    public static String sha256(String text) {
        if (text == null) {
            return "";
        }
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(text.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (Exception ignored) {
            return "";
        }
    }
}

File: python/ptaf/ai/audit/ai_generation_audit_logger.py
from __future__ import annotations

import hashlib
import json
from dataclasses import asdict, dataclass
from pathlib import Path

from ptaf.ai.audit.ai_generation_audit_record import AiGenerationAuditRecord


@dataclass(frozen=True)
class WriteResult:
    written: bool
    output_path: str
    warning_message: str | None


class AiGenerationAuditLogger:
    def append(
        self,
        project_root: Path,
        enabled: bool,
        output_path: str,
        record: AiGenerationAuditRecord,
    ) -> WriteResult:
        if not enabled:
            return WriteResult(False, output_path, None)
        try:
            file_path = (project_root / output_path).resolve()
            file_path.parent.mkdir(parents=True, exist_ok=True)
            with file_path.open("a", encoding="utf-8") as handle:
                handle.write(json.dumps(asdict(record)) + "\n")
            return WriteResult(True, str(file_path), None)
        except Exception:
            return WriteResult(False, output_path, "Warning: audit log could not be written")

    @staticmethod
    def sha256(text: str | None) -> str:
        if text is None:
            return ""
        return hashlib.sha256(text.encode("utf-8")).hexdigest()
